Add a SSE backend

jeremycline opened the pull request:
The SSE backend now uses a Twisted-compatible connection.

Note this is still just a work-in-progress, but I figured I'd put up there for @skrzepto to look at if he wants.

skrzepto committed:
Adds a backend for SSE

This is the first pass at a SSE backend.

jeremycline committed:
The SSE backend now uses a Twisted-compatible connection

Although this introduces a bit of code duplication, it is small and this
change is mostly about reorganizing and consolidating the SSE backend.
This does not include unit tests yet, as I /really/ want to get Twisted
endpoints working with the Twisted connection, which provides connection
manangement.

Signed-off-by: Jeremy Cline <jeremy@jcline.org>

jeremycline commented on fedmsg.d/fas_credentials.py:
```diff
@@ -1,8 +1,6 @@
-import getpass
```
This broke running the tests from the repo's root directory since it'd trigger a password prompt.

pypingou replied:
But doesn't it now breaks running the app itself?

Before you would launch the app, it would ask you for your username and password and work. Now we have to edit this file first or see it crash. Could there be a way to specify a file for the tests?

jeremycline replied:
That's true. On the other hand, if you try to run this as a service, it's going to cause the service to hang forever while it waits for input that's never going to arrive. It's also going to cause a password prompt for every service, even if the service doesn't need the password.

I'd prefer to document that this is a configuration you need to set, and ensure the crash is very obviously because the configuration is missing a username and password.

pypingou replied:
But you shouldn't/wouldn't run this as a service with this file. If you want to run it as a service then you should populate the file, imho, this file is really only there for dev.

jeremycline replied:
The first thing I did when I got started was to copy the config to ~/.fedmsg.d and write a user unit file to run it under systemd, and I hit this. Maybe I shouldn't have, but I did.

I can put it back and either hack around it in the unit tests (which I don't think will be easy) or make a note that you can't run the unit tests from the project root, but I don't like that solution. To me, it's incredibly weird to have code that blocks for user input being executed as part of an `import` statement. Every set of tests is also going to have to account for this, so we're _still_ going to have to document something about this.

jeremycline commented:
I've filed https://github.com/fedora-infra/fmn.consumer/issues/95 to track the Python 3 compatibility issues revealed by the tests.

jeremycline commented:
I've updated this to use the default exchange for publishing so it should be compatible with https://github.com/fedora-infra/fmn.sse/pull/60

jeremycline commented on fmn/consumer/backends/sse.py:
```diff
@@ -0,0 +1,242 @@
+        """
+        short_message = self._format_message(msg, recipient)
+        user = recipient['user']
+        # TODO figure out when it's a group message
```
@pypingou This was one thing I wasn't certain about. The old code just did users. What does a group message look like, and how should this determine whether the message is for a group or a user?

skrzepto replied:
@jeremycline I dont think we have groups set up yet it was just a future feature to implement in fmn. At least thats what I'm aware of.

pypingou replied:
That is correct, it's not there yet but we should have it at one point.

We may want to have another endpoint for groups to make things easier maybe /users/ vs /groups/?

skrzepto replied:
I guess the question now is how do we handle the two endpoints in rabbitmq? Should we just name the queues like `user-<fedora username>` and `group-<fedora groupname>` ?

jeremycline replied:
The way I've got it set up now follows that scheme on the SSE side.

pypingou commented:
We may need to adjust the documentation to mention that fas_credential should be manually updated before running the app.

Otherwise, code-wise, this looks good to me

jeremycline commented:
I've added a documentation project to fmn (https://github.com/fedora-infra/fmn/pull/139) that notes users need to populate ~/.fedmsg.d/fas_credentials.py. I can also add a note to the message of the day.

Edit: Apparently I already added the note in the MOTD: https://github.com/fedora-infra/fmn/blob/develop/ansible/roles/fmn-dev/files/motd#L4

pypingou commented:
Cool :)

jeremycline committed:
Fix the flake8 error causing the tests to fail

Signed-off-by: Jeremy Cline <jeremy@jcline.org>
